rllib/env/single_agent_env_runner.py

```diff
@@ -277,27 +277,34 @@ def _sample(
         force_reset: bool = False,
     ) -> List[SingleAgentEpisode]:
         """Helper method to sample n timesteps or m episodes."""
-
+        ts = 0
+        eps = 0
         done_episodes_to_return: List[SingleAgentEpisode] = []
 
-        # Have to reset the env (on all vector sub_envs).
-        if force_reset or num_episodes is not None or self._needs_initial_reset:
-            episodes = self._episodes = [None for _ in range(self.num_envs)]
-            shared_data = self._shared_data = {}
-            self._reset_envs(episodes, shared_data, explore)
+        def _reset_envs_and_episodes():
+            """Helper method to reset the envs, ongoing episodes and shared data.
+
+            This resets the global ts variable and deletes ongoing episodes.
+            The Done episodes are preserved.
+            """
+            nonlocal ts
+            ts = 0
+            self._ongoing_episodes = [None for _ in range(self.num_envs)]
+            self._shared_data = {}
+            self._reset_envs(self._ongoing_episodes, self._shared_data, explore)
             # We just reset the env. Don't have to force this again in the next
             # call to `self._sample_timesteps()`.
             self._needs_initial_reset = False
```

Comment:
Seems a bit unusual. Any specific reason why this is a nested function?
Maybe we can make it a private method in EnvRunner class? Only diff between single agent and multi agent variants are:
* `ts` -> `env_ts`
* multi agent has `agent_ts`.

Maybe specify type in the method signature, and return all what's needed for the reset? 
We can also get rid of `nonlocal` keyword.

WDYT?